### ISSUE TITLE
test_psa_crypto_config_accel_hash_use_psa is too slow

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#6662]
`test_psa_crypto_config_accel_hash_use_psa` has been timing out often recently.

The goal of this task is to investigate and fix. Should it be split? Can we make it faster somehow? Can we sensibly reduce the scope of what it does?

**Comment by mpg** [comment]
> Can we sensibly reduce the scope of what it does?

I don't think reducing the scope would make a lot of sense. The point of this component is to test that _everything_ works when hashes are provided by only drivers, just as well as it works with software hashes. So, I think testing everything (test suites, ssl-opt.sh, compat.sh) really is what we want here.

We could obviously split into, for example test suites + ssl-opt.sh in one component and compat.sh in another, but I wouldn't find it very satisfying. It ought to be possible to build and run all the tests for a given config in single component. I mean except the fancy config and build, this component is not doing more than `component_test_full_cmake_gcc_asan`...

So, in the long run, I think I'd really like to understand what makes it slow, and if possible make it less slow. But perhaps in the short term we'll have to just split if for the time being.

**Comment by mpg** [comment]
> I do find it weird that it's timing out more than the big test-everything-with-asan component. Dispatching as a driver shouldn't affect performance.

Agreed. I found something that may explain that: the big test-everything-with-asan component uses cmake, which as `-O3` for ASan. This component uses make, and we overwrite the default CFLAGS (which consists of just `-O2`) with our own stuff that doesn't contain any `-O`: so we end up with an non-optimized Asan build, which is quite slow.

This is true of all components that use make and `ASAN_CFLAGS` (except one that uses `$ASAN_CFLAGS -O`, but this one is the first component in that category that runs everything: other run at most `make test` and `ssl-opt.sh -f Default`. (Which btw doesn't really make sense these days: in the past we used `ssl-opt.sh -f Default` to check that a simple handshake worked, but since we now have test in `test_suite_ssl` doing a full handshake, this now seems redundant. This is another occurrence of a long list of "things changed and we never updated out tests".)

Anyway, I tried changing this component to build with `-O2` and the running time on my laptop went from 26m17s to 20m51s, so about 20% less, which is about 12 min less if we start from a running time close to 1h.

Now I'm wondering it we really need to build libtestdriver1 with ASan too. (Clearly I see the value of ASan when building the main libraries, but for this helper library I'm not so sure, as it's only a tool used to test the main libraries with drivers.)

**Comment by mpg** [comment]
Just tried building with ASan only for the main libraries, not libtestdriver1, gained another 8%, so a total 27% which would be a good 15 min starting from about 1h, so I think it's worth it. See https://github.com/Mbed-TLS/mbedtls/pull/6683

**Comment by gilles-peskine-arm** [comment]
I think we've collectively not been paying attention that `ASAN_CFLAGS` is “just the flags for ASan+UBSan” and not “good flags for ASan builds”. How about adding `-O2` to `ASAN_CFLAGS`?

**Comment by mpg** [comment]
> How about adding `-O2` to `ASAN_CFLAGS`?

I've been thinking the same, see #6683.

**Comment by mpg** [comment]
> I'm not convinced that it's useful to run SSL tests with accelerators. What can go wrong when there are accelerators, that is detected when all algorithms are enabled?

I think there are two answers to that. The first is, we want to then analyze the outcomes file and compare to the reference component and ensure that not test was skipped in the driver build that wasn't already skipped in the reference build. This ensures the dependency declarations in `ssl-opt.sh` are correct, but also that we have feature parity between dirver-based and reference build. This reason currently only applies to `ssl-opt.sh`, though, while waiting for `compat.sh` to gain support for outcomes files.

The second reason is, I find it very easy to imagine that we might make mistakes in TLS code by not checking for availability of hashes correctly in all places (like, using `MBEDTLS_SHAxxx_C` in a code path where we'll actually use PSA, either because it's TLS 1.3 or because `USE_PSA` is on, but we made a mistake and used the wrong macro, perhaps because we're far from the place the hash is actually being computed, so we didn't notice). Some of these mistakes might only be apparent when using a certain ciphersuite. Currently the easiest way to ensure all ciphersuites are tested is `compat.sh`.

However, as I write this up, I realise I'm not sure we actually need interop testing with OpenSSL and GnuTLS, so perhaps `compat.sh -p mbedTLS` is enough, and perhaps we can even add `-V YES` to save some time?

Generally speaking, running "all the tests" seemed like it was the easiest way to make sure we didn't miss something, and didn't seem entirely unreasonable at the time. Of course now we're under more pressure to save CI resources, it makes sense to revisit that choice indeed. At the very least, the motivations for running the tests we run should be explained as comments in the component.

**Comment by mpg** [comment]
Looking back at the PR that introduced testing with:
- `ssl-opt.sh` with TLS 1.2 only https://github.com/Mbed-TLS/mbedtls/pull/6255
- `ssl-opt.sh` with TLS 1.3 https://github.com/Mbed-TLS/mbedtls/pull/6256
- `compat.sh` https://github.com/Mbed-TLS/mbedtls/pull/6216
we can see that the PR that introduce `compat.sh` literally did just that, and didn't need to change anything else to make it work, as apparently `compat.sh` didn't uncover any new issue.

So, that suggests that running `compat.sh` might indeed not be that useful here (once we already have a full `ssl-opt.sh` passing).

**Comment by gilles-peskine-arm** [comment]
> Currently the easiest way to ensure all ciphersuites are tested is compat.sh.

Indeed. Soon it'll be done in the unit tests (https://github.com/Mbed-TLS/mbedtls/pull/6645 + follow-up to ensure all cipher suites are covered), but right now it needs `compat.sh`.

> However, as I write this up, I realise I'm not sure we actually need interop testing with OpenSSL and GnuTLS, so perhaps compat.sh -p mbedTLS is enough, and perhaps we can even add -V YES to save some time

Yes, and yes.

**Comment by mpg** [comment]
> Soon it'll be done in the unit tests (#6645 + follow-up to ensure all cipher suites are covered), but right now it needs `compat.sh`.

We should probably make a note that once this is done, we can stop invoking `compat.sh` from this component. (Also, once 6645 is done, and perhaps even now already, I think we can remove all invocations of `ssl-opt.sh -f Default` which are/will be redundant with `test_suite_ssl`.)

**Comment by gilles-peskine-arm** [comment]
https://github.com/Mbed-TLS/mbedtls/pull/6683 has made `component_test_psa_crypto_config_accel_hash_use_psa` significantly faster and I don't remember it being a problem ever since, so I think this issue is resolved.
